src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -60,128 +50,121 @@
+public class CompressionHeaderFactory implements CramCompression {
     private static final Log log = Log.getInstance(CompressionHeaderFactory.class);
-    private static final int oqz = ReadTag.nameType3BytesToInt("OQ", 'Z');
-    private static final int bqz = ReadTag.nameType3BytesToInt("BQ", 'Z');
 
-    public CompressionHeader build(final List<CramCompressionRecord> records, final SubstitutionMatrix substitutionMatrix, final boolean sorted) {
+    private final Map<Integer, EncodingDetails> bestEncodings = new HashMap<Integer, EncodingDetails>();
+
+    /**
+     * Decides on compression methods to use for the given records.
+     *
+     * @param records            the data to be compressed
+     * @param substitutionMatrix a matrix of base substitution frequencies, can be null, in which case it is re-calculated.
+     * @param sorted             if true the records are assumed to be sorted by alignment
+     *                           position
+     * @return {@link htsjdk.samtools.cram.structure.CompressionHeader} object describing the encoding chosen for the data
+     */
+    @Override
+    public CompressionHeader buildCompressionHeader(final List<CramCompressionRecord> records,
+                                                    final SubstitutionMatrix substitutionMatrix, final boolean sorted) {
         final CompressionHeader header = new CompressionHeader();
         header.externalIds = new ArrayList<Integer>();
         int exCounter = 0;
 
         final int baseID = exCounter++;
         header.externalIds.add(baseID);
-        header.externalCompressors.put(baseID,
-                ExternalCompressor.createRANS(RANS.ORDER.ONE));
+        header.externalCompressors.put(baseID, ExternalCompressor.createRANS(RANS.ORDER.ONE));
 
         final int qualityScoreID = exCounter++;
         header.externalIds.add(qualityScoreID);
-        header.externalCompressors.put(qualityScoreID,
-                ExternalCompressor.createRANS(RANS.ORDER.ONE));
+        header.externalCompressors.put(qualityScoreID, ExternalCompressor.createRANS(RANS.ORDER.ONE));
 
         final int readNameID = exCounter++;
         header.externalIds.add(readNameID);
         header.externalCompressors.put(readNameID, ExternalCompressor.createGZIP());
 
-        final int mateInfoID = exCounter++;
-        header.externalIds.add(mateInfoID);
-        header.externalCompressors.put(mateInfoID,
-                ExternalCompressor.createRANS(RANS.ORDER.ONE));
-
         header.encodingMap = new TreeMap<EncodingKey, EncodingParams>();
-        for (final EncodingKey key : EncodingKey.values())
+        for (final EncodingKey key : EncodingKey.values()) {
             header.encodingMap.put(key, NullEncoding.toParam());
+        }
 
         header.tMap = new TreeMap<Integer, EncodingParams>();
 
         { // bit flags encoding:
-            getOptimalIntegerEncoding(header, EncodingKey.BF_BitFlags, 0, records);
+            header.externalIds.add(exCounter);
```

Comment:
It seems like these blocks should be refactored to use a new method on compression header, something like:

`CompressionHeader.addExternalCompressor(EncodingKey encodingKey, ExternalCompressor externalCompressor` 

There's t log of duplicated code here.